services/movie_session.py

```
@@ -0,0 +1,69 @@
+from django.db.models import QuerySet
+from db.models import MovieSession
+from datetime import datetime
+from typing import Optional
+
+
+def create_movie_session(
+        movie_show_time: datetime,
+        movie_id: int,
+        cinema_hall_id: int
+) -> MovieSession:
+    return MovieSession.objects.create(
+        show_time=movie_show_time,
+        movie_id=movie_id,
+        cinema_hall_id=cinema_hall_id
+    )
+
+
+def get_movies_sessions(
+        session_date: Optional[str] = None
+) -> QuerySet:
+    sessions = MovieSession.objects.all()
+    if session_date:
+        try:
+            parsed_date = datetime.strptime(
+                session_date,
+                "%Y-%m-%d"
+            ).date()
+            sessions = sessions.filter(
+                show_time__date=parsed_date
+            )
+        except ValueError:
+            return MovieSession.objects.none()
+    return sessions
+
+
+def get_movie_session_by_id(
+        movie_session_id: int
+) -> Optional[MovieSession]:
+    return MovieSession.objects.get(id=movie_session_id)
+
+
+def update_movie_session(
+        session_id: int,
+        show_time: Optional[datetime] = None,
+        movie_id: Optional[int] = None,
+        cinema_hall_id: Optional[int] = None
+) -> Optional[MovieSession]:
+    session = MovieSession.objects.get(id=session_id)
+    if not session:

```

Logic error: after calling `MovieSession.objects.get(id=session_id)` you check `if not session:`. This is unreachable because `.get()` raises `DoesNotExist` when the object is missing. Either catch `MovieSession.DoesNotExist` around the `.get()` call if you want to return None, or remove this conditional and let the exception propagate. See update_movie_session (line with `.get()`) in services/movie_session.py .